modules/navigation/3d/nav_mesh_queries_3d.cpp

```diff
@@ -127,87 +128,222 @@ Vector3 NavMeshQueries3D::polygons_get_random_point(const LocalVector<gd::Polygo
-			distance_to_point = point.distance_to(p_destination);
-			if (distance_to_point < end_d) {
-				end_d = distance_to_point;
-				end_poly = &p;
-				end_point = point;
-			}
+	if (query_task.callback.is_valid()) {
+		if (emit_callback(query_task.callback)) {
+			query_task.status = NavMeshPathQueryTask3D::TaskStatus::CALLBACK_DISPATCHED;
+		} else {
+			query_task.status = NavMeshPathQueryTask3D::TaskStatus::CALLBACK_FAILED;
 		}
 	}
+}
+
+void NavMeshQueries3D::query_task_polygons_get_path(NavMeshPathQueryTask3D &p_query_task, const LocalVector<gd::Polygon> &p_polygons, const Vector3 &p_map_up, uint32_t p_link_polygons_size) {
+	p_query_task.path_points.clear();
+	p_query_task.path_meta_point_types.clear();
+	p_query_task.path_meta_point_rids.clear();
+	p_query_task.path_meta_point_owners.clear();
+
+	// Find begin polyon and begin position closest to start position and
+	// end polyon and end position closest to target position on the map.
+	const gd::Polygon *begin_poly = nullptr;
+	const gd::Polygon *end_poly = nullptr;
+	Vector3 begin_point;
+	Vector3 end_point;
+
+	_query_task_find_start_end_positions(p_query_task, p_polygons, &begin_poly, begin_point, &end_poly, end_point);
 
 	// Check for trivial cases
 	if (!begin_poly || !end_poly) {
-		return Vector<Vector3>();
+		p_query_task.status = NavMeshPathQueryTask3D::TaskStatus::QUERY_FAILED;
+		return;
 	}
+
 	if (begin_poly == end_poly) {
-		if (r_path_types) {
-			r_path_types->resize(2);
-			r_path_types->write[0] = begin_poly->owner->get_type();
-			r_path_types->write[1] = end_poly->owner->get_type();
-		}
+		_query_task_create_same_polygon_two_point_path(p_query_task, begin_poly, begin_point, end_poly, end_point);
+		return;
+	}
 
-		if (r_path_rids) {
-			r_path_rids->resize(2);
-			(*r_path_rids)[0] = begin_poly->owner->get_self();
-			(*r_path_rids)[1] = end_poly->owner->get_self();
-		}
+	_query_task_build_path_corridor(p_query_task, p_polygons, p_map_up, p_link_polygons_size, begin_poly, begin_point, end_poly, end_point);
 
-		if (r_path_owners) {
-			r_path_owners->resize(2);
-			r_path_owners->write[0] = begin_poly->owner->get_owner_id();
-			r_path_owners->write[1] = end_poly->owner->get_owner_id();
-		}
+	// Post-Process path.
+	if (p_query_task.path_postprocessing == PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL) {
+		_path_corridor_post_process_corridorfunnel(p_query_task, p_query_task.least_cost_id, begin_poly, begin_point, end_poly, end_point, p_map_up);
+	} else if (p_query_task.path_postprocessing == PathPostProcessing::PATH_POSTPROCESSING_EDGECENTERED) {
+		_path_corridor_post_process_edgecentered(p_query_task, p_query_task.least_cost_id, begin_poly, begin_point, end_poly, end_point);
+	} else if (p_query_task.path_postprocessing == PathPostProcessing::PATH_POSTPROCESSING_NONE) {
+		_path_corridor_post_process_nopostprocessing(p_query_task, p_query_task.least_cost_id, begin_poly, begin_point, end_poly, end_point);
+	}
+
+	p_query_task.path_points.invert();
```

Comment:
The path points at this end stage of the pathbuilder are all arranged from target to start so we need to invert them all because users need the path from start to target.